packages/dds/tree/src/test/feature-libraries/editableTree.spec.ts

```diff
@@ -0,0 +1,496 @@
+    schemaMap.set(named.name, named);
+}
+
+const rootPersonSchema = fieldSchema(FieldKinds.value, [personSchema.name]);
+
+const fullSchemaData: SchemaData = {
+    treeSchema: schemaMap,
+    globalFieldSchema: new Map([[rootFieldKey, rootPersonSchema]]),
+};
+
+// TODO: derive types like these from those schema, which subset EditableTree
+
+type Int32 = Brand<number, "Int32">;
+const newAge: Int32 = brand(55);
+
+type ComplexPhoneType = EditableTree & {
+    number: string;
+    prefix: string;
+};
+
+type AddressType = EditableTree & {
+    street: string;
+    zip: string;
+    phones: (number | string | ComplexPhoneType)[];
+};
+
+type PersonType = EditableTree & {
+    name: string;
+    age: Int32;
+    salary: number;
+    friends: Record<string, string>;
+    address: AddressType;
+};
+
+const person: JsonableTree = {
+    type: personSchema.name,
+    fields: {
+        name: [{ value: "Adam", type: stringSchema.name }],
+        age: [{ value: 35, type: int32Schema.name }],
+        salary: [{ value: 10420.2, type: float32Schema.name }],
+        friends: [{ fields: {
+            Mat: [{ type: stringSchema.name, value: "Mat" }],
+        }, type: mapStringSchema.name }],
+        address: [{
+            fields: {
+                street: [{ value: "treeStreet", type: stringSchema.name }],
+                // TODO: revisit as ideally we don't want to have undefined properties in our proxy object
+                zip: [{ type: stringSchema.name }],
+                phones: [{
+                    type: phonesSchema.name,
+                    fields: {
+                        [EmptyKey]: [
+                            { type: stringSchema.name, value: "+49123456778" },
+                            { type: int32Schema.name, value: 123456879 },
+                            { type: complexPhoneSchema.name, fields: {
+                                number: [{ value: "012345", type: stringSchema.name }],
+                                prefix: [{ value: "0123", type: stringSchema.name }],
+                            } },
+                        ],
+                    },
+                }],
+            },
+            type: addressSchema.name,
+        }],
+    },
+};
+
+function setupForest(schema: SchemaData, data: JsonableTree[]): IEditableForest {
+    const schemaRepo = new StoredSchemaRepository(defaultSchemaPolicy, schema);
+    const forest = buildForest(schemaRepo);
+    initializeForest(forest, data);
+    return forest;
+}
+
+function buildTestProxy(data: JsonableTree): UnwrappedEditableField {
+    const forest = setupForest(fullSchemaData, [data]);
+    const [context, field] = getEditableTree(forest);
+    return field;
+}
+
+function buildTestPerson(): PersonType {
+    const proxy = buildTestProxy(person);
+    return proxy as PersonType;
+}
+
+function expectTreeEquals(inputField: UnwrappedEditableField, expected: JsonableTree): void {
+    assert(inputField !== undefined);
+    assert(!Array.isArray(inputField));
```

Comment:
This function takes in an `UnwrappedEditableField`. An `UnwrappedEditableField` explicitly allows `readonly UnwrappedEditableTree[]`. Also, since currently some trees (those with primary sequence fields) get unwrapped into a `readonly UnwrappedEditableTree[]`, it makes sense to allows arrays here.

We have a test which covers this case "array nodes get unwrapped". Its implemented using assert.deepStrictEqual instead of this comparison function since it ended up being slightly simpler / more clear that way, but it should be possible to use expectTreeEquals in its implementation. For example the empty case in that test should be able to be changed to:

```
        // Empty
        {
            const data = { type: phonesSchema.name };
            const forest = setupForest(schemaData, [data]);
            const [context, field] = getEditableTree(forest);
            assert.deepStrictEqual(field, []);
            expectTreeEquals(field, data);
            context.free();
        }
```

and still pass, but now it would fail.

Reply:
Got your point, thanks. Fixed.